### PR TITLE
added shopping cart

### PR DESCRIPTION
**Pull request by TarasYurkovskyi**

### CODE REVIEW

**Review by annton-shakur — APPROVED** [review]
Good job!

File: src/main/java/com/example/bookstore/dto/cart/CartItemRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @NotNull
+    private Long bookId;
+    @Min(value = 1)
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/dto/cart/CartItemRequestDto.java]:
```suggestion
    @Positive
```

File: src/main/java/com/example/bookstore/dto/cart/UpdateCartItemDto.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.example.bookstore.dto.cart;
+
+import jakarta.validation.constraints.Min;
+import lombok.Data;
+
+@Data
+public class UpdateCartItemDto {
+    @Min(value = 1)
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/dto/cart/UpdateCartItemDto.java]:
```suggestion
    @Positive
```

File: src/main/java/com/example/bookstore/repository/ShoppingCartRepository.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import com.example.bookstore.model.ShoppingCart;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart,Long> {
+    @Query("SELECT sc FROM ShoppingCart sc JOIN FETCH sc.cartItems")
+    ShoppingCart findByEmail(String email);
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/repository/ShoppingCartRepository.java]:
```suggestion
    Optional<ShoppingCart> findByEmail(String email);
```

Comment (d1sam) [reply src/main/java/com/example/bookstore/repository/ShoppingCartRepository.java]: should return `Optional` when we write `find` method.

File: src/main/java/com/example/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        CartItem cartItem = cartItemMapper.toEntity(requestDto);
+        Book bookById = bookRepository.findById(requestDto.getBookId()).orElseThrow(
+                () -> new EntityNotFoundException("Cannot get book by id: " + id)
+        );
+
+        cartItem.setBook(bookById);
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(id).orElseThrow(
+                () -> new EntityNotFoundException("Cannot get shopping cart by id: " + id)
+        );
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
           ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseGet(() -> {
      ShoppingCart newShoppingCart = new ShoppingCart();
      newShoppingCart.setUser(user);
      shoppingCartRepository.save(newShoppingCart);
      return newShoppingCart;
    });
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but there are some minor changes should be done!

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**